src/backend/distributed/utils/resource_lock.c

```diff
@@ -244,12 +247,28 @@ lock_shard_resources(PG_FUNCTION_ARGS)
 static void
 LockShardListResourcesOnFirstWorker(LOCKMODE lockmode, List *shardIntervalList)
 {
+	if (!AllowModificationsFromWorkersToReplicatedTables)
+	{
+		/*
+		 * Allowing modifications from worker nodes for replicated tables requires
+		 * to serialize modifications, see AcquireExecutorShardLocksForExecution()
+		 * for the details.
+		 *
+		 * If the user is opted for disabling modifications from the workers, we do not
+		 * need to acquire these remote locks. Returning early saves us from an additional
+		 * network round-trip.
+		 */
+		Assert(IsCoordinator());
```

Comment:
don't workers call this function as well?

Reply:
Ah, of course. Re-run the tests with asserts enabled, and adjusted one more place